source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -124,6 +133,21 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
   return {action, bytes_read, end_stream};
 }
 
+void SslSocket::complete(Envoy::Ssl::PrivateKeyOperationStatus status) {
+  if (status == Envoy::Ssl::PrivateKeyOperationStatus::Success) {
```

Comment:
What kind of threading guarantees does BoringSSL or OpenSSL provide? Do we know that the thread that private key operations are run on is the same as the invoking thread? Should `complete` post to the dispatcher to continue, or is that the job of the extension? This is one place that I'm not 100% sure of the thread correctness.

Reply:
The BoringSSL operations are driven by the caller: BoringSSL's private key complete function will be called when the `SSL_do_handshake()` is called the second time. However, the `complete()` function (indicating that the asynchronous operation is ready and `ssl_socket.cc` needs to call the handshake again) might in principle be called from whatever thread the private key operations extension happens to use. I believe the correct semantics is to define this to be an error, and expect the extension to call the `complete()` function from the worker thread context.

Should we add `ASSERT(isThreadSafe());` to `SslSocket::complete()` the way it is done in `dispatcher_impl.cc`?

I just wrote a (rough) implementation of a QAT private key extension to see if it fits the private key API. There I'm using a pipe and dispatcher to make sure that the callback thread is correct, because QAT device polling is done from a different thread. See the code here: https://github.com/ipuustin/envoy/blob/qat-provider/source/extensions/private_key_operations_providers/qat/qat_private_key_provider.cc#L20

Reply:
Yeah, I think it should be an error to `SslSocket::complete()` on a different thread. The Envoy pattern for handling this is to do a post to the original thread's dispatcher with the completion as the body. The only cross-thread safe dispatcher operation is `post()`.